CRM/Price/BAO/PriceSet.php

```diff
@@ -1747,6 +1765,205 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
+          AND pf.is_active = 1
+    LEFT JOIN civicrm_price_set ps
+           ON ps.id = pf.price_set_id
+          AND ps.is_active = 1
+ WHERE m.id = 34
+       ORDER BY c.receive_date DESC, p.contribution_id DESC
+    ";
+
+    $params = array(1 => array($membership_id, 'Integer'));
+    $dao = self::executeQuery($sql, $params);
+    $activateByDefault = TRUE;
+    while ($dao->fetch()) {
+      // last price set used on this membership (could be multiple over the course of time)
+      // only select "price set view", if last contribution was for the price set.
+      if ($dao->price_set_id) {
+        return array(
+          'price_set_id' => $dao->price_set_id,
+          'price_set_is_through_contribution' => $activateByDefault,
+        );
+      }
+      $activateByDefault = FALSE;
+    }
+
+    // nothing found.  we want to pick, by default the price set that will cover
+    // the most of this contact's members of organizations.
+    $sql = "
+SELECT ps.id, group_concat(DISTINCT member_of_contact_id) AS member_of_contact_ids
+ FROM civicrm_price_set ps
+   INNER JOIN civicrm_price_field pf ON pf.price_set_id = ps.id AND pf.is_active = 1
+   INNER JOIN civicrm_price_field_value pfv ON pfv.price_field_id = pf.id AND pfv.is_active = 1
+   INNER JOIN civicrm_membership_type mt ON mt.id = pfv.membership_type_id
+ WHERE ps.is_active = 1
+ GROUP BY ps.id
+ ";
+
+    $dao = self::executeQuery($sql);
+    $orgsByPriceSet = array();
+    while ($dao->fetch()) {
+      $orgsByPriceSet[$dao->id] = explode(",", $dao->member_of_contact_ids);
+    }
+
+    $memberhsip_type_id = CRM_Member_BAO_Membership::getFieldValue('CRM_Member_BAO_Membership', $membership_id, 'membership_type_id');
+    $member_of_contact_id = CRM_Member_BAO_MembershipType::getFieldValue('CRM_Member_BAO_MembershipType', $memberhsip_type_id, 'member_of_contact_id');
+    $contact_id = CRM_Member_BAO_Membership::getFieldValue('CRM_Member_BAO_Membership', $membership_id, 'contact_id');
+
+    // we've got an array, above, of all orgs mapped on price set id.
+    // now get an array of all orgs for this contact, and return the price set that has the most matches.
+    $contact_membership_orgs = CRM_Member_BAO_Membership::getActiveContactMemberships($contact_id);
+
+    $bestCnt = 0;
+    $bestSet = NULL;
+    foreach ($orgsByPriceSet as $priceSetid => $orgsInPriceSet) {
+      if (!in_array($member_of_contact_id, $orgsInPriceSet)) {
+        continue;
+      }
+
+      $common = count(array_intersect($orgsInPriceSet, $contact_membership_orgs));
+      if ($common > $bestCnt) {
+        $bestCnt = $common;
+        $bestSet = $priceSetid;
+      }
+    }
+
+    return array(
+      'price_set_id' => $bestSet,
+      'price_set_is_through_contribution' => FALSE,
+    );
+  }
+
+  public static function setPriceSetDefaultsToLastUsedValues(&$defaults, $priceSet, $contactID) {
+
+    // this is detailed info on how membership orgs, and the best membership type & price set field to use
```

Comment:
The comment was more specific to the first line.  I missed the doc block part.  I've got that coming now.